lib/cdo/delete_accounts_helper.rb

```diff
@@ -143,28 +139,37 @@ def remove_user_from_sections_as_student(user)
     Follower.with_deleted.where(student_user: user).each(&:really_destroy!)
   end
 
-  # Removes all information about the user pertaining to Pardot. This encompasses Pardot itself, the
-  # contact_rollups pegasus table (master and reporting), and the contact_rollups_daily pegasus table
-  # (reporting only).
-  # @param [Integer] The user ID to purge from Pardot.
-  def remove_from_pardot(user_id)
+  def remove_contacts(email)
+    @pegasus_db[:contacts].where(email: email).delete
+  end
+
+  def remove_from_pardot_and_contact_rollups(contact_rollups_recordset)
+    # TODO: Make this an operation handled by the contact rollups task itself
+    #       instead of crossing the architectural boundary ourselves.
+    #       For now this is unsafe to run while contact rollups is itself running.
     # Though we have the DB tables in all environments, we only sync data from the production
     # environment with Pardot.
     if rack_env == :production
-      pardot_ids = @pegasus_db[:contact_rollups].
+      pardot_ids = contact_rollups_recordset.
```

Comment:
I believe this recordset, in the legitimate case, can only ever have one record? (A single userid or email address.) If so, can we enforce that? Either by just passing in a single Pardot ID to this function, or by throwing if recordset count > 1. Since this deletes data, we want to be really careful and have defense in depth. For example, right now a bug could get introduced in our calling code where we miss a where clause (similar to a recent bug that actually happened in prod), which could pass in a recordset of all contact rollups and delete them all in Pardot.

Reply:
I see what you mean.  I'm concerned that in the situation where our records _are_ malformed, we have some regulatory obligation to delete all rows that might contain PII for the user in question.  In particular, this is going to get more complicated as we separate out the account-delete case from the email-delete case, which are going to follow different rules.

Maybe what needs to happen is a pre-flight validation system that doesn't delete anything unless all affected records appear to be well-formed.  Then if everything looks good we proceed with deletes.  If not, maybe the record goes into a queue for manual review - allowing an engineer to decide "oh, two contact rows, that's fine" vs "oops, we don't want to delete 10,000 contact rows."